### DISCUSSION TITLE
PID Controller Questions

### DISCUSSION THREAD

**Discussion by weberbox** [nebhead/PiFire#34] [Q&A]
So I finally got this project all assembled and installed. My first tests did not go as planned so maybe I am not understanding how the PID part is supposed to work. 

All functions are working from what I can tell and smoke mode seems to do just fine and runs like the normal smoke mode on the old controller. The smoke plus feature is awesome and really does create a lot more smoke. 

When putting PiFire into Hold mode it does not seem to be able to hold lower temps ~180 and seems to run too hot and this also happens at a higher temp of 225 on hold. Ambient was hot today around 90f so maybe that has a little bit to do with it??? 

Setting the hold to 180 the grill never really even comes close. It passes 180 all the way up to ~240. Since I still need to do more learning on PID tuning I have not changed the PB, Ti, Td yet but did change the cycle time all the way up to 90 and still it has a hard time maintaining the temp but does do better than the default 20. 

It will dump more pellets every 90 seconds (or whatever I set for cycle time) and when trying to maintain a lower temp like 180 it seems the auger is on for way too long then temps spike up and never go down. If I set it to Smoke+ it seems to help maintain better temps but still is off by about ~20f and gradually raises over time. 

I am going to start reading about the PID and how to tune it but any advice in the mean time?

**Comment by nebhead** [comment]
OK, so from my experience so far, you may see these issues with the default configuration.  I am no expert with PID theory, and the configuration was borrowed (with permission) from @dborello and his [pismoker](https://github.com/DBorello/PiSmoker) project.  I suspect there may be some tuning that may be required for specific grill sizes, etc.  Here are some of my observations (some of this may be purely anecdotal):

1. The behavior of the grill is dependent in some way on the ambient temperature.  For instance, in the winter time where temperatures drop below 50F outside, the grill will sometimes struggle to get to 140F in smoke mode.  In the summer time, my minimum temperature is somewhere in the 200F range, in smoke or in hold.  The grill can't seem to go below this temperature for whatever reason.  Summer time temps here are regularly in the high 90's F or over 100F.  I had a similar problem with my Traeger controller, so I kind of chalked it up to the default settings being temperature dependent.  I believe this is why they have a "P-Mode" setting for Traeger grills that will adjust the smoke cycle timings.  
2. The wind-up for the PID will almost always overshoot significantly (again with default settings).  However, over time (30 mins or so) the PID will eventually home in on set temperature ranges.  However, if the grill just wants to go to 200F+ minimum temps, you probably won't be going under those (per the observation from number 1).  
3. If you open the lid, it basically resets the PID and it will have to home in on the temperature again which again can take some time.  
4. I also see a dependency on some of the wood types that I use and how hot they will get.  Ultimately, many of the pellets are mostly alder wood, but some have higher concentrations of other woods and may burn hotter or colder.  This can also affect the _minimum_ temperature that you may see.  

Things you might consider trying are to vary the cycle time (sounds like you already moved that to 90).  I'd play with something more conservative like 30 or 40s.  Give that a try and see if that helps.  Pushing to 90 may be burning too much wood, and not giving the PID time to really calibrate.  

Also consider trying to change the derivative time (Td).  Again, I'm no expert, but try to reduce this a bit and see what happens.  

Good luck, let us know how your experimentation goes and hope this helps!

**Comment by weberbox** [comment]
Thanks for the info and help.

Well for # 1 I kind of figured that was playing into the issue. With it being ~95ish outside I imagine that it would be hard to only go up 80-90 and maintain that. I really want to figure out how to circumvent this as slow cooking at 180 in the summer is kind of important to me. Maybe a combination of PID tuning and smoke mode will help. I will need to waste some pellets and find out lol.  

For # 2 just so I am clear will the PID mode always push the auger at the cycle time but just vary the amount of auger on time? I started with 30 then went to 40 then 60 then 90 to try and get it so keep a lower temp. As I went up it seemed to help with the lower temps. 

For # 3 I noticed that the stock pitboss RTD is pretty slow to change and quite a difference since it is on the left side. I was also using a thermoworks air probe at grill level and it was 30+ in difference and would change quite quickly if the lid was opened. Originally I wanted to use the thermoworks at grill level to get a more accurate temp but I might need to use the stock RTD at least in the summer I guess. 

For # 5 I have not played with a lot of pellets and I was thinking about this. Maybe this can help with # 1 if I can find a lower temp burning wood. 

I need to get some more pellets and run some tests. Will report back what I find.

Also I was thinking about the shutdown time. Right now it is about 60 seconds does that work out well for you? I think the stock timer was like 3-4 minutes but I never actually timed it.

**Reply by nebhead** [reply]
> Well for # 1 I kind of figured that was playing into the issue. With it being ~95ish outside I imagine that it would be hard to only go up 80-90 and maintain that. I really want to figure out how to circumvent this as slow cooking at 180 in the summer is kind of important to me. Maybe a combination of PID tuning and smoke mode will help. I will need to waste some pellets and find out lol.

I think it is definitely possible, but will require lots of experimentation, time, patience and pellets. :)  

> For # 2 just so I am clear will the PID mode always push the auger at the cycle time but just vary the amount of auger on time? I started with 30 then went to 40 then 60 then 90 to try and get it so keep a lower temp. As I went up it seemed to help with the lower temps.

Correct.  The design (lifted from pismoker) is that in hold mode (PID) it will always cycle the auger on and off for some amount of time.  The minimum and maximum times are defined in the settings.json file:
```
	settings['cycle_data'] = {
		'PB' : 60.0,
		'Ti' : 180.0,
		'Td' : 45.0,
		'HoldCycleTime' : 20,
		'SmokeCycleTime' : 15,
		'PMode' : 2,  # http://tipsforbbq.com/Definition/Traeger-P-Setting
		'u_min' : 0.15,
		'u_max' : 1.0
	}
 ```
Where `u_min` is the minimum cycle ratio (basically auger on for 15 percent of the time) and `u_max` is auger on for 100% of the time.  

The cycle times are calculated as follows:

```
		CycleRatio = PIDControl.update(AvgGT)
		CycleRatio = max(CycleRatio, settings['cycle_data']['u_min'])
		CycleRatio = min(CycleRatio, settings['cycle_data']['u_max'])
		OnTime = settings['cycle_data']['HoldCycleTime'] * CycleRatio
		OffTime = settings['cycle_data']['HoldCycleTime'] * (1 - CycleRatio)
```

So you _could_ play with the u_min and u_max times as well by editing the settings file directly.  This could yield some interesting results.  

> Also I was thinking about the shutdown time. Right now it is about 60 seconds does that work out well for you? I think the stock timer was like 3-4 minutes but I never actually timed it.

You know, my original Traeger controller did not have a shutdown time at all, so I was kind of just guessing at what a good amount of shutdown time would be.  So far it's been working OK for me, but I haven't really checked the firepot to see if I have a ton of unburned pellets.  This can definitely be bumped up more.  I should probably add a setting for this.  :)  Perhaps that is my next minor update.

**Reply by weberbox** [reply]
So I wanted to look into the PID controller and see if there could be some improvement. It seems to work pretty well once it hones in but that takes a while and it seems to always overshoot by quite a bit. 

During some research I came across this project and it looks like it could be useful. It has the option to pause the PID calculation, use PonM which seems like it could really help the initial overshoot. I am still trying to fully wrap my head around PID and how it works but this blog seems to have a decent amount of info. 

[PID Blog regarding PonM](http://brettbeauregard.com/blog/2017/06/introducing-proportional-on-measurement/) 

[PID Blog](http://brettbeauregard.com/blog/2011/04/improving-the-beginners-pid-introduction/)

[Simple PID Python Project](https://github.com/m-lundberg/simple-pid)

I need to do more reading but I wanted to start playing with this. How cool would it be if the PID could be paused when the lid is opened using a reed switch, proximity sensor or maybe even a distance sensor. And not having to overshoot the initial set temp would be another great benefit.

**Reply by weberbox** [reply]
Ok so I read a bit more on PID and went over the current PID code as well as the PID I linked above. It seems that the current PID already has limits but they are set in control.py by u_min and u_max. The pause might still be cool to toy with but currently I was unable to see a way to improve the current PID. I am going to keep studying and hopefully I will have an epiphany some day. 

During my studies I made this excel file to help understand the PID better as well as have a visual way to play with the PB, Ti, Td, HoldCycleTime and U settings without having to burn pellers. It did make me realize that my higher cycle time was the cause of some of my run away temps since the auger time will be 50% of the HoldCycleTime once the set temp is met. It is not perfect but it really helped me understand how the auger time is calculated. Anyway figured I would share it, maybe it can help someone else.

[PIDControl.ods](https://github.com/nebhead/PiFire/files/7356112/PIDControl.ods)

**Reply by nebhead** [reply]
> Ok so I read a bit more on PID and went over the current PID code as well as the PID I linked above. It seems that the current PID already has limits but they are set in control.py by u_min and u_max. The pause might still be cool to toy with but currently I was unable to see a way to improve the current PID. I am going to keep studying and hopefully I will have an epiphany some day.

Keep at it!  It would be great to see if we can improve upon anything at this point.  I did find another PID controller on GitHub that was very similar, and wanted to try, but didn't really have a chance yet.  But I think you are basically right, that it's all about the inputs into the PID.  

> During my studies I made this excel file to help understand the PID better as well as have a visual way to play with the PB, Ti, Td, HoldCycleTime and U settings without having to burn pellers. It did make me realize that my higher cycle time was the cause of some of my run away temps since the auger time will be 50% of the HoldCycleTime once the set temp is met. It is not perfect but it really helped me understand how the auger time is calculated. Anyway figured I would share it, maybe it can help someone else.

Very cool!  I'll have to also dive into this and see if I can learn something.  Thanks for putting this together!

**Comment by weberbox** [comment]
Ok so I wanted to discuss an option here that I have been playing with.  After using my grill for a while I have this strange feeling it is feeding pellets at a faster rate then Treager or maybe the pellets I am using are just burning hotter. I have been trying to play with the cycle time to help but it seems that at 20 it is just feeding to often when the grill hits temp which over time causes the temp to gradually rise over the set point. When raising cycle time to allow a longer gap of feeding it seems to improve a bit but the PID can feed even more pellets as it is homing in. 

I made a setting to change the 0.5 ratio in pid.py this way as it nears or is at the set point the amount of auger on time can be reduced. Currently from my testing the only thing that has really affected auger time when very close to or at the set point is cycle time. 

I need to do more testing but it is behaving much better when attempting to meet and hold the temp. In my case it also requires some tweaking of cycle time, min cycle ratio and max cycle ratio. I am currently testing with a 25 cycle time, 0.1 min cycle ratio, 0.75 max cycle ratio and a center ratio of 0.25 vs the hard coded 0.5

Anyway I wanted to ask if you would like me to make a pull request to add the option to PiFire. I called it center ratio as that was all I could think of at the time. I can change that if you have a better idea. 

```
		settings = ReadSettings()
		self.Center = settings['cycle_data']['center']

		self.Derv = 0.0
		self.Inter = 0.0
		self.Inter_max = abs(self.Center/self.Ki)

		self.Last = 150

		self.setTarget(0.0)

	def CalculateGains(self,PB,Ti,Td):
		self.Kp = -1/PB
		self.Ki = self.Kp/Ti
		self.Kd = self.Kp*Td

	def update(self, Current):
		#P
		error = Current - self.setPoint
		self.P = self.Kp*error + self.Center
```

**Reply by nebhead** [reply]
Absolutely!  I've observed the same basic behavior and I'm relatively certain that it burns more pellets than your standard Traeger controller might.  I just haven't had the guts to experiment with the PID settings to try and optimize anything.  I'd love it if you could make a pull-request for this so I can also tweak these settings.  I think there may be some variability in ambient temperature and pellet types, but in general I think it is a step in the right direction.

**Reply by weberbox** [reply]
> Absolutely! I've observed the same basic behavior and I'm relatively certain that it burns more pellets than your standard Traeger controller might. I just haven't had the guts to experiment with the PID settings to try and optimize anything. I'd love it if you could make a pull-request for this so I can also tweak these settings. I think there may be some variability in ambient temperature and pellet types, but in general I think it is a step in the right direction.

Ok I will get it sent over with the other updates or shortly after. With all these options it can take a lot of tuning but I think that is a good thing since there are many variables at play. 

One other idea I had was possibly a way to store and restore the pid settings along with the pellets. Maybe it is overkill but since some pellets burn hotter or cooler it would be pretty cool to store the pid settings per pellets then auto load them when you change pellets in the pellet manager, but I wonder how many would use this or take the time to tune per pellet brand/wood.

**Reply by nebhead** [reply]
> One other idea I had was possibly a way to store and restore the pid settings along with the pellets. Maybe it is overkill but since some pellets burn hotter or cooler it would be pretty cool to store the pid settings per pellets then auto load them when you change pellets in the pellet manager, but I wonder how many would use this or take the time to tune per pellet brand/wood.

Oh that's a super cool idea!  I'll take a look at implementing something like that after you push the update.  

I've been toying with the idea of having a GitHub repo for probe profiles, etc. that would allow folks to download / submit profiles.  Might be more trouble than it's worth.  Having the ability to download pellet profiles that are community tuned would be super neat.

**Reply by weberbox** [reply]
Well I did a cook yesterday with the new PID settings and for me it was a great improvement. The overshoot of temp was way lower than before and once it hit the 225 I set it never crept up and held within about 5f unless I opened the lid etc. and even then it recovered much better. I think with some more tuning of max cycle I might be able to lower the overshoot even more. 

Also ran this cook with the new Celsius/Fahrenheit update and all went fine. Although I did not run the grill in C it did work just fine in F. 

Cycle Time - 25
Min Cycle - 0.1
Max Cycle - 0.75
Center Ratio - 0.25
PB Ti, Td were default.

![Screenshot from 2021-12-12 18-16-28](https://user-images.githubusercontent.com/313328/145915380-f19a84d7-0e89-4949-8ca2-653d9294af40.png)

**Reply by nebhead** [reply]
> Well I did a cook yesterday with the new PID settings and for me it was a great improvement. The overshoot of temp was way lower than before and once it hit the 225 I set it never crept up and held within about 5f unless I opened the lid etc. and even then it recovered much better. I think with some more tuning of max cycle I might be able to lower the overshoot even more.

Awesome!  I also did a cook over the weekend (some pork chops that were already in the sous vide bath for about 2 hours), and it went OK for me with the new settings.  However, I did notice a couple things.  First, my grill really struggles to hit even 300F when the ambient temperatures are around 50F.  And with the new tuning, the auger was running constantly...  and now I am realizing that I'm an idiot.  I didn't reset the settings to the new PID settings.  Well that's embarrassing.  OK, I will go update those settings and try it again with next cook and report back.  Oops. :)

**Reply by weberbox** [reply]
Haha happens to the best of us. I hope you get some good results next time your try. I think each grill is probably a little different including what pellets are used etc. I did not push mine past the 225 for this cook but it hit about 250 without issue when the overshoot happened and it was like 51f outside. For me the auger ran less due to the .25 center and .75 max cycle. When it hits temp and settles the auger cycles are really short, I was almost worried of the pot going out but it had no issues. I think the great thing is now that almost everything on the PID can be controlled via the UI or app it will allow everyone to tune a little closer even with different grills, auger speeds, pellets etc.

**Comment by thomasstrach** [comment]
Hi,

maybe I missed the essential part:

What changes do I have to make to get my PiFire running in Celsius or are you still working on it?

 

Von: James Weber ***@***.***> 
Gesendet: Dienstag, 14. Dezember 2021 02:30
An: nebhead/PiFire ***@***.***>
Cc: Subscribed ***@***.***>
Betreff: Re: [nebhead/PiFire] PID Controller Questions (Discussion #34)

 

Well I did a cook yesterday with the new PID settings and for me it was a great improvement. The overshoot of temp was way lower than before and once it hit the 225 I set it never crept up and held within about 5f unless I opened the lid etc. and even then it recovered much better. I think with some more tuning of max cycle I might be able to lower the overshoot even more.

Also ran this cook with the new Celsius/Fahrenheit update and all went fine. Although I did not run the grill in C it did work just fine in F.

Cycle Time - 25
Min Cycle - 0.1
Max Cycle - 0.75
Center Ratio - 0.25
PB Ti, Td where default.

 <https://user-images.githubusercontent.com/313328/145915380-f19a84d7-0e89-4949-8ca2-653d9294af40.png>

**Reply by weberbox** [reply]
Celsius is supported but currently only with the development branch. It probably needs some testing as well to verify everything works as expected unless nebhead has already done that. I tested the branch and it seemed to be ok but have not run an actual cook with Celsius yet.

**Reply by nebhead** [reply]
> Hi, maybe I missed the essential part: What changes do I have to make to get my PiFire running in Celsius or are you still working on it? 

Yeah, the Celsius changes are now in the main branch so you should be able to pull the latest.  You can change to Celsius in settings or if installing from scratch, in the installer.  It's been relatively well tested, but do let us know if there are any issues.

**Comment by weberbox** [comment]
Just wanted to show this. It was a cook from last night. With the new PID settings this thing is amazing. 2-4f swing with the lid closed and the over/undershoot was cut down so much for my grill. 

![Screenshot from 2022-02-07 11-35-12](https://user-images.githubusercontent.com/313328/152859062-41c84f56-9313-4b59-8518-2ccd2b8866d4.png)

**Comment by FastEddie49** [comment]
Picking up the discussion from the Discord channel...does anyone know what the conditions are for PiFire to begin using the PID calcs? A recent cook of mine showed that PID is only used for 44% of the auger ON time...the remainder of the total auger ON time is spent within the MinCycle/MaxCycle range. My thought is not to rework or even understand the PID algo but to understand the conditions that cause the program to use the PID and possibly come up with a way to utilize the PID calculations more often. Also, it kinda nags at me that no provision is made for auger feed rate. I know this isn't an exact precise thing, feeding pellets, and precise temperature control is likely not possible nor required...it's just that I always thought PID was a full-time proposition and in this instance it's not. I think it would be ideal if it were, but that may not be possible...ok, how do we influence the PID to be in control more often? Can that be done? What are the conditions that trigger the PID calcs to be used?

**Reply by nebhead** [reply]
My perspective is a little different, so bear with me here.  I would argue that the PID isn't 'off' ever.  It's just giving us a value that is outside of our min-max cycle range.  For example, this is a real cook that I did recently that shows the difference between the set-point temp and the actual grill temp (this is the Error for the PID)(blue line) and shows the cycle ratio that PiFire is using during these periods.  And yes, the cycle ratio will hit maximums and minimums frequently when either the Error goes positive or negative too far.  But when we start to get closer with errors to the setpoint, its giving you more in-between cycle ratios. 

![image](https://user-images.githubusercontent.com/19876886/202321213-81a2bb4a-433f-4b6f-88ae-711c6bc33b7d.png)

So the PID is _always in control_.  And in fact it does its job pretty well, by centering the temperature around the setpoint.  Could it be better?  Maybe.  But as discussed, might require a PhD in control theory to fully grasp. :P
